File: 2024/day02_Red-Nosed_Reports/main.ts
import * as fs from 'fs';

(() => {
    type Report = number[];
    type Reports = Report[];
    type Results = {
        safeCt: number;
        safeCtWithDampener: number;
    };

    const readInput = (filename: string): Reports =>
        fs.readFileSync(filename, 'utf8')
            .trim()
            .split(/\r?\n/)
            .map(line => line.split(/\s+/).map(Number));

    const isSafe = (report: Report): boolean => {
        const deltas = report.slice(1).map((value, i) => value - report[i]);
        const normalizedDeltas = deltas[0] < 0 ? deltas.map(d => -d) : deltas;
        return normalizedDeltas.every(d => d >= 1 && d <= 3);
    };

    const solve = (reports: Reports): Results => {
        let safeCt = 0;
        let safeCtWithDampener = 0;

        for (const report of reports) {
            if (isSafe(report)) {
                safeCt++;
            } else {
                for (let i = 0; i < report.length; i++) {
                    const dampenedReport = [...report.slice(0, i), ...report.slice(i + 1)];
                    if (isSafe(dampenedReport)) {
                        safeCtWithDampener++;
                        break;
                    }
                }
            }
        }

        return { safeCt: safeCt, safeCtWithDampener: safeCt + safeCtWithDampener };
    };

    const main = () => {
        const filename = process.argv[2] || 'input.txt';
        const reports = readInput(filename);
        const { safeCt, safeCtWithDampener } = solve(reports);

        console.log("Part 1: safe_ct =", safeCt);
        console.log("Part 2: safe_ct_with_dampener =", safeCtWithDampener);
    };

    main();
})();

File: 2024/day02_red-nosed_reports/main.rs
#![feature(array_windows)]
use std::{env, fs};

fn is_safe(report: &[i32]) -> bool {
    let mut c = 1;
    if report[1] - report[0] < 0 {
        c = -1;
    }
    for i in 1..report.len() {
        let v = c * (report[i] - report[i-1]);
        if v < 1 || v > 3 {
            return false
        }
    }
    true
}

fn solve(reports: &Vec<Vec<i32>>) -> (i32, i32) {
    let mut safe_ct = 0;
    let mut safe_ct_with_dampening = 0;
    for report in reports {
        if is_safe(&report) {
            safe_ct += 1;
        } else {
            for i in 0..report.len() {
                let mut dampened_report = report.clone();
                dampened_report.remove(i);
                if is_safe(&dampened_report) {
                    safe_ct_with_dampening += 1;
                    break;
                }
            }
        }
    }
    (safe_ct, safe_ct + safe_ct_with_dampening)
}
fn read_input(filename: &str) -> Vec<Vec<i32>> {
    fs::read_to_string(filename).expect("Unable to open file")
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|v| v.parse::<i32>().unwrap())
                .collect::<Vec<_>>()
        }).collect()
}

fn main() {
    let mut filename = "input.txt";
    let args: Vec<String> = env::args().collect();
    if args.len() == 2 {
        filename = &args[1];
    }

    let reports = read_input(filename);
    let (safe_ct, safe_ct_with_dampening) = solve(&reports);
    println!("Part1: safe_ct = {}", safe_ct);
    println!("Part2: safe_ct_with_dampening = {}", safe_ct_with_dampening);
}
